prov/rxm/src/rxm_cq.c

```diff
@@ -999,12 +1000,16 @@ static ssize_t rxm_cq_handle_comp(struct rxm_ep *rxm_ep,
 
 	switch (RXM_GET_PROTO_STATE(comp->op_context)) {
 	case RXM_TX:
-	case RXM_INJECT_TX:
 		tx_eager_buf = comp->op_context;
 		assert(comp->flags & FI_SEND);
 		ret = rxm_finish_eager_send(rxm_ep, tx_eager_buf);
 		rxm_tx_buf_release(rxm_ep, RXM_BUF_POOL_TX, tx_eager_buf);
 		return ret;
+	case RXM_INJECT_TX:
+		tx_base_buf = comp->op_context;
+		assert(comp->flags & FI_SEND);
+		rxm_tx_buf_release(rxm_ep, RXM_BUF_POOL_TX_INJECT, tx_base_buf);
```

Comment:
What is the full impact of this bug?  Do the buffer pools maintain similar structures, such that one can be re-used, and we're really only looking at a memory leak?  Or is this something worse?

Reply:
So, one of the differences between the two pools is that buffers from RXM_BUF_POOL_TX_INJECT wouldn't have hdr field set. I'm surprised we haven't seen a segfault when functions using buffers from RXM_BUF_POOL_TX try to access the hdr.

Reply:
I *just* fixed a similar problem in the tcp provider, though the likelihood of hitting that issue would have been rare.  This is obviously something that we need to try to detect. It looks like we should have this ability already with a debug build, so I'll try to add an appropriate assert.

We should try to figure out why we aren't seeing constant crashes without this change.

Reply:
It looks like we don't need this fix at all. Buffers from RXM_BUF_POOL_TX_INJECT are used only for fi_inject calls and that would never generate a completion from MSG provider. So we would never hit this switch case.

Also, the buffers are returned back to the right pool after fi_inject succeeds. So there wouldn't any sort of memory leak.

I'll drop the patch.

Reply:
Does that mean the location of 'case RXM_INJECT_TX' is correct, or that we shouldn't actually hit it?  Should this fall through to the default case or have an assertion?

Reply:
We shouldn't hit the case. I'll add an assertion that the state is not RXM_INJECT_TX when we get a completion.

Also, it seems almost all of the fabtests set FI_THREAD_DOMAIN. We would use buffers from the inject pool only when it's FI_THREAD_SAFE. We're missing some test coverage there. I think fi_ubertest doesn't have a threading level config as well.

Reply:
fi_ubertest should test FI_THREAD_SAFE as that's the default threading level would be returned. I dropped this commit and added a new one that adds the assert.